File: 416. Partition Equal Subset Sum/[space optimize]416. Partition Equal Subset Sum .go
func canPartition(nums []int) bool {
    sum := 0
    for _,num := range nums {
        sum += num
    }
    if sum & 1 == 1 {
        return false
    }
    s := int(sum/2)
    dp := make([]bool,s+1)
    dp[0] = true
    for _,num := range nums{
        for i := s; i>0;i-- {
            if i>=num {
                dp[i] = dp[i] || dp[i-num]
            }
        }
    }
    return dp[s]
}

File: 416. Partition Equal Subset Sum/[space optimize]416. Partition Equal Subset Sum .py
# since each element should only be used once, the second for loop should in a desc order.
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        s = sum(nums)
        if s&1 == 1:
            return False
        s //=2
        n = len(nums)
        dp = [False for _ in range(s+1)]
        dp[0] = True
        for i in range(n):
            for j in range(s,0,-1):
                if j>=nums[i]:
                    dp[j] = dp[j] or dp[j-nums[i]]
        return dp[s]
